Ensure ctx.fireChannelReadComplete() is only called when something was d...

normanmaurer opened the pull request:
...ecoded

Motivation:

ctx.fireChannelReadComplete() should only be called if something is decoded during a channelRead(...) operation. Also we must ensure that it will be called
if channelRead(...) decodes something at some point as channelRead(...) maybe called multiple times by the transport before channelReadComplete(...) is called.

Modifications:

- Ensure channelReadComplete(...) only triggers ctx.fireChannelReadComplete() when a previous channelRead(...) call decoded a message.
- Fix a bug where ctx.read() was not called if the last channelRead(...) operation did not decode a message but previous did.

Result:

Correct semantic with channelReadComplete(...) events and also ensure no stales

normanmaurer commented:
@trustin please review again.

trustin commented on codec-http/src/main/java/io/netty/handler/codec/spdy/SpdyFrameCodec.java:
```diff
@@ -96,6 +97,15 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
     @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+        if (read) {
+            // something was read we should notify the read of handlers in the pipeline.
```
Please focus on punctuation. :-)

Something was read. We should notify the next handlers in the pipeline.

trustin commented on codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java:
```diff
@@ -176,7 +176,12 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
+                if (size > 0 && decodeWasNull) {
+                    // channelRead(...) may be called multiple times before channelReadComplete(...) so we need
+                    // ensure that if we decoded at least one message decodeWasNull is set so channelReadComplete(...)
+                    // handles it correctly.
```
I cannot parse this comment. Punctuation?

trustin commented on codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java:
```diff
@@ -212,8 +217,10 @@ public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
                 ctx.read();
             }
+        } else {
+            // Something was read, call fireChannelReadComplete()
```
Something was read. Trigger channelReadComplete()

trustin commented on codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:
```diff
@@ -160,4 +163,61 @@ public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+    // See https://github.com/netty/netty/pull/3263
+    @Test
+    public void testFireChannelReadCompleteWhenDecodeOnce() {
+        final AtomicBoolean readComplete = new AtomicBoolean(false);
```
Please use `AtomicInteger` just in case `channelReadComplete()` is invoked twice (or more.)

trustin commented:
Please address the comments and ping me for another round.

normanmaurer commented:
@trustin done.. please check

normanmaurer commented:
@trustin ping...

trustin commented:
I wonder if we can implement this in our `ChannelHandlerContext` implementation. What do you think?

normanmaurer commented:
You mean by surpress events even if the user fire them by hand? 



> Am 20.01.2015 um 06:53 schrieb Trustin Lee <notifications@github.com>:
> 
> I wonder if we can implement this in our ChannelHandlerContext implementation. What do you think?
> 
> —
> Reply to this email directly or view it on GitHub.
>

trustin commented:
Yeah, just suppress `channelReadComplete()` without `channelRead()`.

normanmaurer commented:
@trustin actually that's a great idea... let me do this and ping you once done

normanmaurer commented:
@trustin while doing the changes I noticed one drawback...

In `ByteToMessageDecoder` we do:

```java
    @Override
    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
        if (cumulation != null && !first && cumulation.refCnt() == 1) {
            // discard some bytes if possible to make more room in the
            // buffer but only if the refCnt == 1  as otherwise the user may have
            // used slice().retain() or duplicate().retain().
            //
            // See:
            // - https://github.com/netty/netty/issues/2327
            // - https://github.com/netty/netty/issues/1764
            cumulation.discardSomeReadBytes();
        }
        if (decodeWasNull) {
            decodeWasNull = false;
            if (!ctx.channel().config().isAutoRead()) {
                ctx.read();
            }
        } else {
            // Something was read. Trigger channelReadComplete().
            ctx.fireChannelReadComplete();
        }
    }
```

This means we trigger read() in this case to ensure we not getting stuck. We can't do this in `AbstractChannelHandlerContext` as the `ChannelHandler` could be the last in the `ChannelPipeline` and so the user may want to not call read() but do it later. What we could is add some "marker interface" or Annotation and have `ByteToMessageDecoder` and `MessageToMessageDecoder` implement / annotated with it. This way we could detect if we should trigger read or not. 

WDYT ?

trustin commented:
I'm not sure I get what you are trying to say.

Note that `ByteToMessageDcoder.channelRead(Complete)()` will be invoked by the previous handler, and the changes we are going to make will not affect the behavior of `ByteToMessageDecoder`.

All you need to change in `AbstractChannelHandlerContext` is,

1. Add a boolean flag (or something similar)
2. set the flag to true in `fireChannelRead()`
3. clear the flag in `fireChannelReadComplete()` if the flag is set.  Otherwise, just suppress the event.

It might get slightly more complicated if a user specifies a different executor or a different invoker for a handler.

`ByteToMessageDecoder.channelReadComplete()` could be simplified a little bit though, but it doesn't really matter here.

normanmaurer commented:
@trustin I'm just saying even if we push the channelReadComplete() stuff into AbstractChannelHandlerContext we will still need to have do the work for call `ChannelHandlerContext.read()` in the handler itself, which means it is stateful. I would prefer to be able to do all of this in `AbstractChannelHandlerContext` (call read if we need read more) as otherwise our MessageToMessageDecoder will have the same issue and need to be stateful. 

@trustin does this make it clear ?

normanmaurer commented:
@trustin one idea to "fix this" would be to check `if next != tail` and if so trigger ctx.read() if nothing was produced + autoRead == false.

So something like this:

```java
    private void invokeNextChannelReadComplete(AbstractChannelHandlerContext next) {
        if (channelReadCompletePending) {
            channelReadCompletePending = false;
            next.invokeChannelReadComplete();
        } else if (next != tail && !channel.config().isAutoRead) {
            // Call read() as this handler was not the last in the pipeline and autoRead is false
            read();
        }
    }
```

normanmaurer commented:
Hey @trustin , so I implemented your suggestion in  a7e2cab and added  70273a1 to show what I meant by my other comment. 

Let me know what you think about both of the commits. Please ignore the commit messages for now. I will come-up with a proper msg if we like it.

trustin commented:
I guess we can't make such an assumption (i.e. if `next != tail`, we are OK to trigger `ctx.read()`), because it's entirely up to a user how the user programs the pipeline. No?

trustin commented:
Basically, my concern is that such a change might trigger an unwanted `ctx.read()`, just in case I wasn't clear.

normanmaurer commented:
@trustin yeah I agree with your concern... I just think it is the best that we can do to make it happen in AbstractChannelHandlerContext. This was the reason why I said we may want to add a marker interface or annotation which we use in ByteToMessageDecoder and MessageToMessageDecoder, so we can detect that we want to trigger it. As in this case we want to trigger a read in all cases (like we do atm in ByteToMessageDecoder and should do in MessageToMessageDecoder)

Hopefully that makes it clear.

normanmaurer commented:
@trustin I just pushed another commit which may fix the "auto read stuff". What I did was just add a protected method to `ChannelInboundHandlerAdapter` that allows a handler to override if read will be triggered automatically if `channelRead(...)` only consumed input. The default is false but `ByteToMessageDecoder` and `MessageToMessageDecoder` override this method and return true. 

I think this should fix the issue and is in line with what we did with `ChannelHandlerAdapter.isSharable()`.

Let me know what you think about it... Also the method name `isTriggerRead()` is not good, so I'm happy for suggestions to replace it with something better.

trustin commented:
I had a chance to think a little bit more about this topic and I think it could be done by modifying `AbstractChannelHandlerContext` only.

First, we need a boolean flag that is set to true when the current handler's `read()` method is invoked.  i.e. When the flag is set, it means the handler in the upstream has called `ctx.read()`.

Second, we need another boolean flag that is set to true when the current handler calls `ctx.fireChannelRead()`.

When the current handler's `channelReadComplete()` method is invoked, the first flag is true, and the second flag is false, we are safe to call `ctx.read()` automatically.

normanmaurer commented:
@trustin please check again... I did the changes to follow your idea. It seems like it works as expected, yay :)

trustin commented:
LGTM. Thanks for your patience, but the elegance of the resulting change should justify it. Please merge. :-)

normanmaurer commented:
@trustin agreed :) Cherry-picked into 4.0 (3e70b4ed99cbd29798a869bb3d7b3415ca8416e0) . For 4.1 and master I'm not sure yet how to fix it best as we moved the invocation logic to the ChannelHandlerInvoker. Will need to think more about it

trustin commented:
Will you also fix this for 4.1 and master? (Just in case we forget about this :-))

normanmaurer commented:
@trustin yep... Just not sure yet how



> Am 02.02.2015 um 09:29 schrieb Trustin Lee <notifications@github.com>:
> 
> Will you also fix this for 4.1 and master? (Just in case we forget about this :-))
> 
> —
> Reply to this email directly or view it on GitHub.
>

trustin commented on transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:
```diff
@@ -311,19 +317,24 @@ private void invokeChannelRead(Object msg) {
     }
 
+    private void invokeNextChannelRead(AbstractChannelHandlerContext next, Object msg) {
+        nextChannelReadInvoked = true;
```
If I understood correctly, `nextChannelReadInvoked` can be updated by any non-I/O thread, because this method is being executed by the potentially different executor than the executor of the current context.

trustin replied:
That is, when we update the pipeline dynamically, thread safety might break.

normanmaurer replied:
Shoildn't this method be executed only by the correct Thread? What I'm missing?



> Am 07.02.2015 um 06:05 schrieb Trustin Lee <notifications@github.com>:
> 
> In transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:
> 
> > @@ -311,19 +317,24 @@ private void invokeChannelRead(Object msg) {
> >          }
> >      }
> >  
> > +    private void invokeNextChannelRead(AbstractChannelHandlerContext next, Object msg) {
> > +        nextChannelReadInvoked = true;
> If I understood correctly, nextChannelReadInvoked can be updated by any non-I/O thread, because this method is being executed by the potentially different executor than the executor of the current context.
> 
> —
> Reply to this email directly or view it on GitHub.
>

trustin replied:
Note that the executor that invokes this method is the next context's executor, which can be anything, especially when a user replaces the next handler with another handler with different executor.

normanmaurer replied:
@trustin hmm... You are working on a fix or should I do?



> Am 07.02.2015 um 07:07 schrieb Trustin Lee <notifications@github.com>:
> 
> In transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:
> 
> > @@ -311,19 +317,24 @@ private void invokeChannelRead(Object msg) {
> >          }
> >      }
> >  
> > +    private void invokeNextChannelRead(AbstractChannelHandlerContext next, Object msg) {
> > +        nextChannelReadInvoked = true;
> Note that the executor that invokes this method is the next context's executor, which can be anything, especially when a user replaces the next handler with another handler with different executor.
> 
> —
> Reply to this email directly or view it on GitHub.
>

trustin replied:
I have a fix which makes the new boolean fields volatile. If that's fine, then I can push the fix quickly.

normanmaurer replied:
Sounds good



> Am 07.02.2015 um 07:23 schrieb Trustin Lee <notifications@github.com>:
> 
> In transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:
> 
> > @@ -311,19 +317,24 @@ private void invokeChannelRead(Object msg) {
> >          }
> >      }
> >  
> > +    private void invokeNextChannelRead(AbstractChannelHandlerContext next, Object msg) {
> > +        nextChannelReadInvoked = true;
> I have a fix which makes the new boolean fields volatile. If that's fine, then I can push the fix quickly.
> 
> —
> Reply to this email directly or view it on GitHub.
>

trustin commented on transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:
```diff
@@ -311,19 +317,24 @@ private void invokeChannelRead(Object msg) {
                     @Override
                     public void run() {
-                        next.invokeChannelReadComplete();
+                        invokeNextChannelReadComplete(next);
```
This is actually not a good idea because the next context will then have a reference to the current context. i.e. we replaced next.invoke*() to this.invokeNext*(). I think this is potentially very difficult to follow?

trustin replied:
And if pipeline manipulation occurs after this task is cached, it will break things.

normanmaurer replied:
Hmm... How isvthis different to what we had before. If I understood you correct the issue existed also before



> Am 07.02.2015 um 07:03 schrieb Trustin Lee <notifications@github.com>:
> 
> In transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:
> 
> >          } else {
> >              Runnable task = next.invokeChannelReadCompleteTask;
> >              if (task == null) {
> >                  next.invokeChannelReadCompleteTask = task = new Runnable() {
> >                      @Override
> >                      public void run() {
> > -                        next.invokeChannelReadComplete();
> > +                        invokeNextChannelReadComplete(next);
> And if pipeline manipulation occurs after this task is cached, it will break things.
> 
> —
> Reply to this email directly or view it on GitHub.
>

trustin replied:
Let's say you cached the next context's invokeChannelReadCompleteTask. The task here will call this.invokeNextChannelReadComplete() where `this` is the previous context. So far so good.

However, if a user removes the previous context from the pipeline, `next`'s previous context is not `this` anymore.  Hence, `invokeChannelReadCompleteTask` will update the flag of the wrong context, which has been removed already.

trustin commented:
I guess the simple yet proper solution is to make the boolean variables volatile and update and check the variables before invoking anything. It also makes it easy to port the change to 4.1 and master.

trustin commented:
Fixed again in 8b2fb2b (4.0), 27a25e2 (4.1), and 14d64d0 (master)
